src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs

```diff
@@ -58,60 +62,129 @@ public class QueryService : IQueryService
 
         var partitionKey = _contextAccessor.RequestContext.GetPartitionKey();
 
+        Stopwatch stopwatch = new Stopwatch();
+
+        stopwatch.Start();
         QueryResult queryResult = await _queryStore.QueryAsync(partitionKey, queryExpression, cancellationToken);
+        stopwatch.Stop();
+        var filterTime = stopwatch.ElapsedMilliseconds;
 
         if (!queryResult.DicomInstances.Any())
         {
             return new QueryResourceResponse(Array.Empty<DicomDataset>(), queryExpression.ErroneousTags);
         }
 
-        var responseBuilder = new QueryResponseBuilder(queryExpression);
-        IEnumerable<DicomDataset> responseMetadata = null;
+        var responseBuilder = new QueryResponseBuilder(queryExpression, _contextAccessor.RequestContext.ApiMajorVersion);
 
-        if (QueryLimit.ContainsComputedTag(queryExpression.IELevel, responseBuilder.ReturnTags))
-        {
-            var versions = queryResult.DicomInstances.Select(i => i.Version).ToList();
+        stopwatch.Restart();
+        IEnumerable<DicomDataset> instanceMetadata = await GetInstanceMetadataAsync(partitionKey, queryExpression, queryResult, responseBuilder.ReturnTags, cancellationToken);
+        stopwatch.Stop();
+        var resultTime = stopwatch.ElapsedMilliseconds;
+        _logger.LogInformation("QueryService performance filterTimeMilliseconds:{FilterTime}, resultTimeMilliseconds:{ResultTime}", filterTime, resultTime);
+
+        var responseMetadata = instanceMetadata.Select(m => responseBuilder.GenerateResponseDataset(m));
+        return new QueryResourceResponse(responseMetadata, queryExpression.ErroneousTags);
+    }
 
-            if (queryExpression.IELevel == Messages.ResourceType.Study)
+    // Does not handle retrieving the extendedQueryTag indexes right now. Logs are in place to evaluate it in the future.
+    private async Task<IEnumerable<DicomDataset>> GetInstanceMetadataAsync(
```

Comment:
This method is pretty dense. Could we walk through it offline?

Reply:
ofcourse, will schedule sometime on Monday